lib/kennel/parts_writer.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require "find"
+
+module Kennel
+  class PartsWriter
+    def initialize(filter:)
+      @filter = filter
+    end
+
+    def write(parts)
+      Progress.progress "Storing" do
+        if filter.tracking_id_filter
+          write_changed(parts)
+        else
+          old = old_paths
+          used = write_changed(parts)
+          (old - used).each { |p| FileUtils.rm_rf(p) }
+        end
+      end
+    end
+
+    private
+
+    attr_reader :filter
+
+    def write_changed(parts)
+      used = ["generated"]
+
+      Utils.parallel(parts, max: 2) do |part|
+        path = "generated/#{part.tracking_id.tr("/", ":").sub(":", "/")}.json"
+        used.concat [File.dirname(path), path] # only 1 level of sub folders, so this is safe
+        payload = part.as_json.merge(api_resource: part.class.api_resource)
+        write_file_if_necessary(path, JSON.pretty_generate(payload) << "\n")
+      end
+
+      used
```

Comment:
since we can add the folder multiple times we should `uniq`
```suggestion
      used.uniq
```

Reply:
Sure. Again, this was a problem with the original code too, but I am happy that refactoring like this is helping bringing these problems to the fore so they can be fixed :)

Reply:
ecb1a4ab81983270ed6e012f63f0d490e279df6b